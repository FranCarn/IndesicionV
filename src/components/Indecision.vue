<template>
  <img v-if="url" :src="url" alt="bg" />
  <div class="bg-dark">
    <div class="indesicion-container">
      <input
        type="text"
        placeholder="Make a quest..."
        v-model="question"
        @keypress.enter="makeQuestion"
      />
      <p>Use question sign's</p>

      <div v-if="isValidCuestion">
        <h2>{{ oldQuest }}</h2>
        <h1>{{ answer }}</h1>
      </div>
    </div>
  </div>
</template>

<script>
export default {
  name: "Indesicion",
  data() {
    return {
      question: "",
      answer: "",
      oldQuest: "",
      url: null,
      isValidCuestion: false,
    };
  },
  methods: {
    async makeQuestion() {
      try {
        this.isValidCuestion = true;
        this.url = null;
        this.answer = "Thinking...";
        this.oldQuest = this.question;
        this.question = "";
        const { image, answer } = await fetch("https://yesno.wtf/api").then(
          (res) => res.json()
        );
        this.url = image;
        this.answer = answer;
      } catch (error) {
        console.log(error);
      }
    },
  },
  watch: {
    question(value, oldValue) {
      this.isValidCuestion = false;
      console.log({ value, oldValue });
      if (!value.includes("?")) return;

      this.makeQuestion();
    },
  },
};
</script>

<style scoped>
img,
.bg-dark {
  height: 100vh;
  left: 0px;
  max-height: 100%;
  max-width: 100%;
  position: fixed;
  top: 0px;
  width: 100vw;
}

.bg-dark {
  background-color: rgba(0, 0, 0, 0.4);
}

.indecision-container {
  position: relative;
  z-index: 99;
}

input {
  width: 250px;
  padding: 10px 15px;
  border-radius: 5px;
  border: none;
  margin-top: 150px;
  margin-bottom: 15px;
}
input:focus {
  outline: none;
}

p {
  color: white;
  font-size: 20px;
  margin-top: 0px;
}

h1,
h2 {
  color: white;
}
h1 {
  text-transform: capitalize;
}
h2 {
  margin-top: 150px;
}
</style>
