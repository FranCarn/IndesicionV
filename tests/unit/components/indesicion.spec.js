import { shallowMount } from "@vue/test-utils";
import Indecision from "@/components/Indecision";

describe("Indesicion component", () => {
  let wrapper;
  let clgSpy;
  let makeQuestion;

  global.fetch = jest.fn(() =>
    Promise.resolve({
      json: () =>
        Promise.resolve({
          image: "https://yesno.wtf/assets/yes/2.gif",
          answer: "yes",
          forced: false,
        }),
    })
  );

  beforeEach(() => {
    wrapper = shallowMount(Indecision);
    clgSpy = jest.spyOn(console, "log");
    makeQuestion = jest.spyOn(wrapper.vm, "makeQuestion");
    jest.clearAllMocks();
  });

  test("should match with snapshop", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
  test("should write in input & dont fire any action", async () => {
    const input = wrapper.find("input");
    await input.setValue("Hi World");
    expect(clgSpy).toHaveBeenCalledTimes(1);
    expect(makeQuestion).not.toHaveBeenCalled();
  });
  test("should fire fetch when question sign is writed", async () => {
    const input = wrapper.find("input");
    await input.setValue("Hi World?");
    expect(makeQuestion).toHaveBeenCalledTimes(1);
  });
  test("test on makeQuestion", async () => {
    await wrapper.vm.makeQuestion();
    const img = wrapper.find("img");
    expect(img.exists()).toBeTruthy();
    expect(wrapper.vm.url).toBe("https://yesno.wtf/assets/yes/2.gif");
    expect(wrapper.vm.answer).toBe("yes");
  });
  test("should fail makeQuestion", async () => {
    fetch.mockImplementationOnce(() => Promise.reject("API down"));
    await wrapper.vm.makeQuestion();
    const img = wrapper.find("img");

    expect(img.exists()).toBeFalsy();
    expect(wrapper.vm.answer).toBe("Cannot connect to API");
  });
});
