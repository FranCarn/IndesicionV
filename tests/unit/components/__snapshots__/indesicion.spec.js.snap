// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indesicion component should match with snapshop 1`] = `
<!--v-if-->
<div class="bg-dark">
  <div class="indesicion-container"><input type="text" placeholder="Make a quest...">
    <p>Use question sign's</p>
    <!--v-if-->
  </div>
</div>
`;
